### PR TITLE
Adding head and count option on select(), insert(), update(), delete(), and rpc()

### PR DESCRIPTION
**Pull request by dshukertjr**
## What kind of change does this PR introduce?

It will add `head` and `count` option to allow querying row count. 

## What is the current behavior?

Cannot get row count. 

## What is the new behavior?

Enable retrieving row count. 

## Related issues

#94

### CODE REVIEW

**Comment by dshukertjr** [comment]
I have implemented `count` option for select, and wanted some feedback so that what I am doing here is along the line of what discussed [here](https://github.com/supabase/postgrest-js/issues/94).

**Comment by dshukertjr** [comment]
@steve-chavez  @kiwicopple  @soedirgo 
I am done implementing the `count` and `head` option on `select`, and added `count` field to the test snapshot at this point. If everything looks fine, I would move on to implementing `count` and `head` to `insert()`, `update()`, `delete()`, and `rpc()`.

File: src/lib/types.ts [hunk 1]
```diff
@@ -80,18 +82,36 @@ export abstract class PostgrestBuilder<T> implements PromiseLike<PostgrestRespon
...
-          const isReturnMinimal = this.headers['Prefer']?.split(',').includes('return=minimal')
-          data = isReturnMinimal ? null : await res.json()
+          if (this.method !== 'HEAD') {
+            const isReturnMinimal = this.headers['Prefer']?.split(',').includes('return=minimal')
+            data = isReturnMinimal ? null : await res.json()
+          } else {
+            data = null
+          }
+
+          const countHeader = this.headers['Prefer']?.match(/count=(exact|planned|estimated)/)
```
Comment (dshukertjr) [inline src/lib/types.ts:95]: I feel like there is a better way of extracting the `count` value from the header here. Any suggestions would be great!

File: src/lib/PostgrestQueryBuilder.ts [hunk 2]
```diff
@@ -17,8 +17,13 @@ export default class PostgrestQueryBuilder<T> extends PostgrestBuilder<T> {
    * Performs vertical filtering with SELECT.
    *
    * @param columns  The columns to retrieve, separated by commas.
+   * @param head  When set to true, select will void data.
+   * @param count  Count algorithm to use to count rows in a table.
    */
-  select(columns = '*'): PostgrestFilterBuilder<T> {
+  select(
+    columns = '*',
+    { head, count }: { head?: boolean; count?: 'exact' | 'planned' | 'estimated' } = {}
```
Comment (soedirgo) [inline src/lib/PostgrestQueryBuilder.ts:25]: Might be better to explicitly set the default values here:
```suggestion
    {
      head = false,
      count = null,
    }: {
      head?: boolean
      count?: null | 'exact' | 'planned' | 'estimated'
    } = {}
```

File: src/lib/types.ts (hunk 1 above)
Comment (soedirgo) [reply src/lib/types.ts:95]: I think this is fine for now. An alternative would be to add a `prefers` object property to `PostgrestBuilder` and join everything in `then()`.

**Review by soedirgo — COMMENTED** [review]
Nice work @dshukertjr, just one nit :)

Also, if you could add a simple test case in `basic.ts` that would be great!

File: src/lib/types.ts (hunk 1 above)
Comment (dshukertjr) [reply src/lib/types.ts:95]: @soedirgo  Okay, I will leave it as it is for now. Thanks for checking!

File: src/lib/PostgrestQueryBuilder.ts (hunk 2 above)
Comment (dshukertjr) [reply src/lib/PostgrestQueryBuilder.ts:25]: Thanks for checking. I will commit your suggestion.

**Comment by dshukertjr** [comment]
I will work on adding the test cases now.

File: test/__snapshots__/index.test.ts.snap [hunk 3]
```diff
@@ -2006,6 +2714,50 @@ Object {
...
+  "error": null,
+  "status": 200,
+  "statusText": "OK",
+}
+`;
+
+exports[`select with head:true, count:estimated 1`] = `
+Object {
+  "body": null,
+  "count": 510,
```
Comment (dshukertjr) [inline test/__snapshots__/index.test.ts.snap:2731]: count: estimated outputs 510 when the actual count is 4. Does anyone have any suggestion about what to do with this?

**Comment by dshukertjr** [comment]
I do have one concern about the result of `count: estimate` and `count: planned`. Other than that, I think this pull request is pretty close to be ready to be merged. Any comments, feedback is greatly welcome!

File: test/__snapshots__/index.test.ts.snap (hunk 3 above)
Comment (dshukertjr) [reply test/__snapshots__/index.test.ts.snap:2731]: I just read that we need to run `ANALYZE` on the table to update the estimate count in the plan. This seems like a separate issue, so I think we should just leave this 510 for right now and maybe deal with it in a separate pull request. What does everybody think? @steve-chavez  @kiwicopple  @soedirgo

Comment (soedirgo) [reply test/__snapshots__/index.test.ts.snap:2731]: Oh, so `estimated` and `planned` just give estimates, so the number can fluctuate sometimes. I think it's fine to either just test for `exact` or just snapshot the `data`.

Comment (soedirgo) [reply test/__snapshots__/index.test.ts.snap:2731]: On second thought, not much point in testing if it doesn't test the count. Let's keep these in for now and hope it behaves...

**Comment by soedirgo** [comment]
Awesome work @dshukertjr, love the extensive tests :)

Will merge this if it looks good @kiwicopple

File: test/__snapshots__/index.test.ts.snap (hunk 3 above)
Comment (dshukertjr) [reply test/__snapshots__/index.test.ts.snap:2731]: Sounds good!

Comment (kiwicopple) [reply test/__snapshots__/index.test.ts.snap:2731]: These numbers will definitely change based on the number of tuples Postgres needs to clean up. Perhaps we can use a [Property Matcher](https://jestjs.io/docs/en/snapshot-testing#property-matchers) to check that it's returning a(ny) number?

Comment (dshukertjr) [reply test/__snapshots__/index.test.ts.snap]: @kiwicopple Great idea. I just updated the test code to do so!

**Comment by dshukertjr** [comment]
Since the last concern that I had with this PR is solved thanks to @kiwicopple , I think this PR is ready to be merged.

**Pull request merged**